{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db47bb03-b787-4cf7-9203-352dc7eb6ecd",
   "metadata": {},
   "source": [
    "The purpose of this notebook is to demonstrate the prototype gene mapper library put together to generalize the processes of\n",
    "\n",
    "- Mapping gene symbols to gene identifiers\n",
    "- Mapping gene identifiers across authorities (NCBI, ENSEMBL, etc.)\n",
    "- Mapping genes in one species to ortholog genes in another species\n",
    "\n",
    "for MapMyCells.\n",
    "\n",
    "In addition to the `mmc_gene_mapper` codebase and its dependencies, this notebook depends on the `abc_atlas_access` tool, though only for downloading metadata to simulate mapping 10s of thousands of genes at once (as we would with real user data)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aa87849-b621-4dff-b1b7-3d1bf8e87995",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "de6c8e20-d464-4151-b975-fe82d4014fec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py:67: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import h5py\n",
    "import json\n",
    "import pathlib\n",
    "import subprocess\n",
    "import sqlite3\n",
    "import time\n",
    "\n",
    "import abc_atlas_access.abc_atlas_cache.abc_project_cache as abc_cache_library\n",
    "\n",
    "import mmc_gene_mapper\n",
    "import mmc_gene_mapper.utils.file_utils as file_utils\n",
    "import mmc_gene_mapper.ensembl_download.scraper as ensembl_scraper\n",
    "import mmc_gene_mapper.mapper.mapper as mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7351d2c5-f2fb-4446-8586-c9c2e25e9947",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/abc_atlas_access/abc_atlas_cache/cloud_cache.py:630: OutdatedManifestWarning: You are loading releases/20250331/manifest.json. A more up to date version of the dataset -- releases/20250531/manifest.json -- exists online. To see the changes between the two versions of the dataset, run\n",
      "type.compare_manifests('releases/20250331/manifest.json', 'releases/20250531/manifest.json')\n",
      "To load another version of the dataset, run\n",
      "type.load_manifest('releases/20250531/manifest.json')\n",
      "  warnings.warn(msg, OutdatedManifestWarning)\n"
     ]
    }
   ],
   "source": [
    "abc_cache = abc_cache_library.AbcProjectCache.from_cache_dir('../data/abc_cache')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "eea118fa-0f34-4027-a06a-28543fbce932",
   "metadata": {},
   "outputs": [],
   "source": [
    "bican_files = [\n",
    "    {'url': 'https://ftp.ensembl.org/pub/release-101/gff3/homo_sapiens/Homo_sapiens.GRCh38.101.gff3.gz',\n",
    "     'assembly_id': 'GCF_000001405.39'},\n",
    "    {'url': 'https://ftp.ensembl.org/pub/release-98/gff3/mus_musculus/Mus_musculus.GRCm38.98.gff3.gz',\n",
    "     'assembly_id': 'GCF_000001635.26'}\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7d6b0ff2-8c0c-45ed-8e0f-bd11688ca775",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "considered 25 dir; 27 entries\n",
      "considered 50 dir; 52 entries\n",
      "considered 75 dir; 77 entries\n",
      "considered 100 dir; 102 entries\n",
      "considered 125 dir; 127 entries\n",
      "considered 150 dir; 152 entries\n",
      "considered 175 dir; 177 entries\n",
      "considered 200 dir; 202 entries\n",
      "considered 225 dir; 227 entries\n",
      "considered 250 dir; 252 entries\n",
      "considered 275 dir; 277 entries\n",
      "considered 300 dir; 302 entries\n",
      "considered 325 dir; 327 entries\n",
      "processing 348 entries\n",
      "processed 34 of 348 in 5.13e+02 seconds\n",
      "processed 68 of 348 in 7.24e+02 seconds\n",
      "processed 102 of 348 in 9.70e+02 seconds\n",
      "    == skipping Homo_sapiens.GRCh38.114.gff3.gz; already serialized\n",
      "processed 136 of 348 in 1.22e+03 seconds\n",
      "processed 170 of 348 in 1.47e+03 seconds\n",
      "processed 204 of 348 in 1.57e+03 seconds\n",
      "processed 238 of 348 in 1.65e+03 seconds\n",
      "processed 272 of 348 in 1.74e+03 seconds\n",
      "processed 306 of 348 in 1.83e+03 seconds\n",
      "processed 340 of 348 in 1.91e+03 seconds\n",
      "failed files\n",
      "{\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/aquila_chrysaetos_chrysaetos/Aquila_chrysaetos_chrysaetos.bAquChr1.2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '223781'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/astyanax_mexicanus_pachon/Astyanax_mexicanus_pachon.Astyanax_mexicanus-1.0.2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/bison_bison_bison/Bison_bison_bison.Bison_UMD1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '43346'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/bos_indicus_hybrid/Bos_indicus_hybrid.UOA_Brahman_1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/bos_taurus_hybrid/Bos_taurus_hybrid.UOA_Angus_1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/buteo_japonicus/Buteo_japonicus.ButJap1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '224669'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '6239'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cairina_moschata_domestica/Cairina_moschata_domestica.CaiMos1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/camarhynchus_parvulus/Camarhynchus_parvulus.Camarhynchus_parvulus_V1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '87175'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/canis_lupus_familiarisbasenji/Canis_lupus_familiarisbasenji.Basenji_breed-1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/canis_lupus_familiarisboxer/Canis_lupus_familiarisboxer.Dog10K_Boxer_Tasha.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/canis_lupus_familiarisgreatdane/Canis_lupus_familiarisgreatdane.UMICH_Zoey_3.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/canis_lupus_familiarisgsd/Canis_lupus_familiarisgsd.UU_Cfam_GSD_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/capra_hircus_blackbengal/Capra_hircus_blackbengal.CVASU_BBG_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/capra_hircus_gca015443085v1/Capra_hircus_gca015443085v1.Saanen_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/capra_hircus_gca026652205v1/Capra_hircus_gca026652205v1.ASM2665220v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/colobus_angolensis_palliatus/Colobus_angolensis_palliatus.Cang.pa_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '336983'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cottoperca_gobio/Cottoperca_gobio.fCotGob3.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '56716'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cricetulus_griseus_chok1gshd/Cricetulus_griseus_chok1gshd.CHOK1GS_HDv1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cricetulus_griseus_crigri/Cricetulus_griseus_crigri.CriGri_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cricetulus_griseus_picr/Cricetulus_griseus_picr.CriGri-PICRH-1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cyprinus_carpio_carpio/Cyprinus_carpio_carpio.Cypcar_WagV4.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '630221'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cyprinus_carpio_germanmirror/Cyprinus_carpio_germanmirror.German_Mirror_carp_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cyprinus_carpio_hebaored/Cyprinus_carpio_hebaored.Hebao_red_carp_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/cyprinus_carpio_huanghe/Cyprinus_carpio_huanghe.Hunaghe_carp_2.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/erythrura_gouldiae/Erythrura_gouldiae.GouldianFinch.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/felis_catus_abyssinian/Felis_catus_abyssinian.Felis_catus_9.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/gadus_morhua_gca010882105v1/Gadus_morhua_gca010882105v1.gadMor_Celtic.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/gallus_gallus_gca000002315v5/Gallus_gallus_gca000002315v5.GRCg6a.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/gallus_gallus_gca016700215v2/Gallus_gallus_gca016700215v2.bGalGal1.pat.whiteleghornlayer.GRCg7w.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/gasterosteus_aculeatus_gca006229185v1/Gasterosteus_aculeatus_gca006229185v1.BOT.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/gasterosteus_aculeatus_gca006232265v1/Gasterosteus_aculeatus_gca006232265v1.BAM.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/gasterosteus_aculeatus_gca006232285v1/Gasterosteus_aculeatus_gca006232285v1.SYL.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/heterocephalus_glaber_female/Heterocephalus_glaber_female.Naked_mole-rat_maternal.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/heterocephalus_glaber_male/Heterocephalus_glaber_male.Naked_mole-rat_paternal.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/homo_sapiens/Homo_sapiens.GRCh38.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 151, in serialize_bkbit_gff3\\n    raise RuntimeError(f\\\"{taxon_id} already serialized\\\")\\nRuntimeError: NCBITaxon:9606 already serialized\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/lonchura_striata_domestica/Lonchura_striata_domestica.LonStrDom1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/malurus_cyaneus_samueli/Malurus_cyaneus_samueli.mCya_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 242, in __init__\\n    self.organism_taxon = self.generate_organism_taxon(taxon_id)\\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 351, in generate_organism_taxon\\n    name=self.taxon_common_name[taxon_id],\\n         ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^\\nKeyError: '2593467'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mesocricetus_auratus/Mesocricetus_auratus.MesAur1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mola_mola/Mola_mola.ASM169857v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/monodelphis_domestica/Monodelphis_domestica.ASM229v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/monodon_monoceros/Monodon_monoceros.NGI_Narwhal_1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/monopterus_albus/Monopterus_albus.M_albus_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/moschus_moschiferus/Moschus_moschiferus.MosMos_v2_BIUU_UCD.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 321, in __download_gff_file\\n    f_gzip.write(data)\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 517, in __exit__\\n    result = self.file.__exit__(exc, value, tb)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_caroli/Mus_caroli.CAROLI_EIJ_v1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp889gepk0.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus/Mus_musculus.GRCm39.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpuiruy1e8.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_129s1svimj/Mus_musculus_129s1svimj.129S1_SvImJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_aj/Mus_musculus_aj.A_J_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_akrj/Mus_musculus_akrj.AKR_J_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_balbcj/Mus_musculus_balbcj.BALB_cJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_c3hhej/Mus_musculus_c3hhej.C3H_HeJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_c57bl6nj/Mus_musculus_c57bl6nj.C57BL_6NJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_casteij/Mus_musculus_casteij.CAST_EiJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_cbaj/Mus_musculus_cbaj.CBA_J_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_dba2j/Mus_musculus_dba2j.DBA_2J_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_fvbnj/Mus_musculus_fvbnj.FVB_NJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_lpj/Mus_musculus_lpj.LP_J_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_molossinusjf1msj/Mus_musculus_molossinusjf1msj.JF1_MsJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_nodshiltj/Mus_musculus_nodshiltj.NOD_ShiLtJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_nzohlltj/Mus_musculus_nzohlltj.NZO_HlLtJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_pwkphj/Mus_musculus_pwkphj.PWK_PhJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_musculus_wsbeij/Mus_musculus_wsbeij.WSB_EiJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_pahari/Mus_pahari.PAHARI_EIJ_v1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_spicilegus/Mus_spicilegus.MUSP714.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mus_spretus/Mus_spretus.SPRET_EiJ_v3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/mustela_putorius_furo/Mustela_putorius_furo.MusPutFur1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/myotis_lucifugus/Myotis_lucifugus.Myoluc2.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 130, in serialize_bkbit_gff3\\n    gff3.parse()\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 532, in parse\\n    f_out.write(f_in.read())\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/myripristis_murdjan/Myripristis_murdjan.fMyrMur1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 321, in __download_gff_file\\n    f_gzip.write(data)\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 517, in __exit__\\n    result = self.file.__exit__(exc, value, tb)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/naja_naja/Naja_naja.Nana_v5.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmptf4ruwtp.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/nannospalax_galili/Nannospalax_galili.S.galili_v1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmptobbzovy.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/neogobius_melanostomus/Neogobius_melanostomus.RGoby_Basel_V2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpiuh65qbv.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/neolamprologus_brichardi/Neolamprologus_brichardi.NeoBri1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmplzqyvmhv.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/neovison_vison/Neovison_vison.NNQGG.v01.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/nomascus_leucogenys/Nomascus_leucogenys.Nleu_3.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp5puj1caf.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/notamacropus_eugenii/Notamacropus_eugenii.Meug_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmppwoe4p8r.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/notechis_scutatus/Notechis_scutatus.TS10Xv2-PRI.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpdefn9mq2.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/nothobranchius_furzeri/Nothobranchius_furzeri.Nfu_20140520.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpap9r3gnn.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/nothoprocta_perdicaria/Nothoprocta_perdicaria.notPer1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpwea9ph_p.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/numida_meleagris/Numida_meleagris.NumMel1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpbw32sget.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ochotona_princeps/Ochotona_princeps.OchPri2.0-Ens.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpsqdt9osr.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/octodon_degus/Octodon_degus.OctDeg1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpqd9p_8vw.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oncorhynchus_kisutch/Oncorhynchus_kisutch.Okis_V2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp_9cddjyd.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oncorhynchus_mykiss/Oncorhynchus_mykiss.USDA_OmykA_1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpiop2dgt1.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oncorhynchus_tshawytscha/Oncorhynchus_tshawytscha.Otsh_v2.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpi5kncfgw.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oreochromis_aureus/Oreochromis_aureus.ZZ_aureus.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp0ptal_5w.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oreochromis_niloticus/Oreochromis_niloticus.O_niloticus_UMD_NMBU.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpdxsgon_m.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ornithorhynchus_anatinus/Ornithorhynchus_anatinus.mOrnAna1.p.v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpuex1bdpe.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oryctolagus_cuniculus/Oryctolagus_cuniculus.OryCun2.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpmw7gyuzx.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oryzias_javanicus/Oryzias_javanicus.OJAV_1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 321, in __download_gff_file\\n    f_gzip.write(data)\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 517, in __exit__\\n    result = self.file.__exit__(exc, value, tb)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oryzias_latipes/Oryzias_latipes.ASM223467v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpn6prjgg6.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oryzias_latipes_hni/Oryzias_latipes_hni.ASM223471v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oryzias_latipes_hsok/Oryzias_latipes_hsok.ASM223469v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oryzias_melastigma/Oryzias_melastigma.Om_v0.7.RACA.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpyh8zb_v6.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/oryzias_sinensis/Oryzias_sinensis.ASM858656v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpye7byzw6.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/otolemur_garnettii/Otolemur_garnettii.OtoGar3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpr27w0ma9.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/otus_sunia/Otus_sunia.OtuSun1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp1gjowwqx.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries/Ovis_aries.Oar_v3.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpsge4bqhc.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca011170295v1/Ovis_aries_gca011170295v1.ASM1117029v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca018804185v1/Ovis_aries_gca018804185v1.NWAFU_Friesian_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca022416685v1/Ovis_aries_gca022416685v1.ASM2241668v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca022416695v1/Ovis_aries_gca022416695v1.ASM2241669v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca022416755v1/Ovis_aries_gca022416755v1.ASM2241675v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca022416775v1/Ovis_aries_gca022416775v1.ASM2241677v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca022416915v1/Ovis_aries_gca022416915v1.ASM2241691v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca022432825v1/Ovis_aries_gca022432825v1.ASM2243282v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca022432835v1/Ovis_aries_gca022432835v1.ASM2243283v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_gca024222175v1/Ovis_aries_gca024222175v1.ASM2422217v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ovis_aries_rambouillet/Ovis_aries_rambouillet.ARS-UI_Ramb_v2.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pan_paniscus/Pan_paniscus.panpan1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 321, in __download_gff_file\\n    f_gzip.write(data)\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 517, in __exit__\\n    result = self.file.__exit__(exc, value, tb)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pan_troglodytes/Pan_troglodytes.Pan_tro_3.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpj5rumz6p.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/panthera_leo/Panthera_leo.PanLeo1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpwv0iaq_y.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/panthera_pardus/Panthera_pardus.PanPar1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpzj1y4p5u.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/panthera_tigris_altaica/Panthera_tigris_altaica.PanTig1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmptmnla_bk.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/papio_anubis/Papio_anubis.Panubis1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpy695iv13.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/parambassis_ranga/Parambassis_ranga.fParRan2.2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpso_0wejk.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/paramormyrops_kingsleyae/Paramormyrops_kingsleyae.PKINGS_0.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp6p9cabo_.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/parus_major/Parus_major.Parus_major1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp9tbhtm90.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pavo_cristatus/Pavo_cristatus.AIIM_Pcri_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpde90hf91.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pelodiscus_sinensis/Pelodiscus_sinensis.PelSin_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmps8n2ge3e.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pelusios_castaneus/Pelusios_castaneus.Pelusios_castaneus-1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpwr7qzius.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/periophthalmus_magnuspinnatus/Periophthalmus_magnuspinnatus.PM.fa.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmph5pwtim2.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/peromyscus_maniculatus_bairdii/Peromyscus_maniculatus_bairdii.HU_Pman_2.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp8uyut78n.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/petromyzon_marinus/Petromyzon_marinus.Pmarinus_7.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp46ya4oso.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/phascolarctos_cinereus/Phascolarctos_cinereus.phaCin_unsw_v4.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpsyw_sfqf.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/phasianus_colchicus/Phasianus_colchicus.ASM414374v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp6zm1dg_z.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/phocoena_sinus/Phocoena_sinus.mPhoSin1.pri.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpj4gzp80p.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/physeter_catodon/Physeter_catodon.ASM283717v2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/piliocolobus_tephrosceles/Piliocolobus_tephrosceles.ASM277652v2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpco2iwqsp.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/podarcis_muralis/Podarcis_muralis.PodMur_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp_6ix2f5c.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/poecilia_formosa/Poecilia_formosa.PoeFor_5.1.2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpotqfjamf.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/poecilia_latipinna/Poecilia_latipinna.P_latipinna-1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp_wm370vi.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/poecilia_mexicana/Poecilia_mexicana.P_mexicana-1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp90y3q6zn.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/poecilia_reticulata/Poecilia_reticulata.Guppy_female_1.0_MT.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp1x5vwx5n.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pogona_vitticeps/Pogona_vitticeps.pvi1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp_o6x2umh.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pongo_abelii/Pongo_abelii.Susie_PABv2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpq5r_cm07.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/procavia_capensis/Procavia_capensis.proCap1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmprlysdt2p.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/prolemur_simus/Prolemur_simus.Prosim_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmphku9c69j.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/propithecus_coquereli/Propithecus_coquereli.Pcoq_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpxgorpek3.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pseudonaja_textilis/Pseudonaja_textilis.EBS10Xv2-PRI.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpyj8gqj7v.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pteropus_vampyrus/Pteropus_vampyrus.pteVam1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpgppevz4z.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pundamilia_nyererei/Pundamilia_nyererei.PunNye1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp4m90gzp_.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/pygocentrus_nattereri/Pygocentrus_nattereri.fPygNat1.pri.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpjh8n01hq.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/rattus_norvegicus/Rattus_norvegicus.GRCr8.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpjm84azod.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/rattus_norvegicus_shrspbbbutx/Rattus_norvegicus_shrspbbbutx.UTH_Rnor_SHRSP_BbbUtx_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/rattus_norvegicus_shrutx/Rattus_norvegicus_shrutx.UTH_Rnor_SHR_Utx.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/rattus_norvegicus_wkybbb/Rattus_norvegicus_wkybbb.UTH_Rnor_WKY_Bbb_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/rhinolophus_ferrumequinum/Rhinolophus_ferrumequinum.mRhiFer1_v1.p.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpaeaemym7.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/rhinopithecus_bieti/Rhinopithecus_bieti.ASM169854v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpws0rm9a2.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/rhinopithecus_roxellana/Rhinopithecus_roxellana.Rrox_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpk1txopbp.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpdxx6n8gu.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/saimiri_boliviensis_boliviensis/Saimiri_boliviensis_boliviensis.SaiBol1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmphz8oh9g0.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/salarias_fasciatus/Salarias_fasciatus.fSalaFa1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmprr2t_bp4.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/salmo_salar/Salmo_salar.Ssal_v3.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpkaihs8u8.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/salmo_salar_gca021399835v1/Salmo_salar_gca021399835v1.USDA_NASsal_1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/salmo_salar_gca923944775v1/Salmo_salar_gca923944775v1.Ssal_Brian_v1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/salmo_salar_gca931346935v2/Salmo_salar_gca931346935v2.Ssal_ALTA.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/salmo_trutta/Salmo_trutta.fSalTru1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp9uyhcvr0.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/salvator_merianae/Salvator_merianae.HLtupMer3.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp3qm_p5p2.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sander_lucioperca/Sander_lucioperca.SLUC_FBN_1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpyb2wrd19.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sarcophilus_harrisii/Sarcophilus_harrisii.mSarHar1.11.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp5gluo1bd.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sciurus_vulgaris/Sciurus_vulgaris.mSciVul1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpgooom_vj.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/scleropages_formosus/Scleropages_formosus.fSclFor1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp842ivxdv.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/scophthalmus_maximus/Scophthalmus_maximus.ASM1334776v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpr8v7x8hj.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/serinus_canaria/Serinus_canaria.SCA1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp0vdqv_a3.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/seriola_dumerili/Seriola_dumerili.Sdu_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpmhnndp58.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/seriola_lalandi_dorsalis/Seriola_lalandi_dorsalis.Sedor1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpbk6bcnk2.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sinocyclocheilus_anshuiensis/Sinocyclocheilus_anshuiensis.SAMN03320099.WGS_v1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpxgwisvou.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sinocyclocheilus_grahami/Sinocyclocheilus_grahami.SAMN03320097.WGS_v1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp37cjrasz.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sinocyclocheilus_rhinocerous/Sinocyclocheilus_rhinocerous.SAMN03320098_v1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpjz826lrs.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sorex_araneus/Sorex_araneus.COMMON_SHREW1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp8c79fgdu.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sparus_aurata/Sparus_aurata.fSpaAur1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpekep81rw.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/spermophilus_dauricus/Spermophilus_dauricus.ASM240643v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpwtqh8ngv.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sphaeramia_orbicularis/Sphaeramia_orbicularis.fSphaOr1.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpp5l0dous.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sphenodon_punctatus/Sphenodon_punctatus.ASM311381v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp5ircqasl.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/stachyris_ruficeps/Stachyris_ruficeps.ASM869450v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/stegastes_partitus/Stegastes_partitus.Stegastes_partitus-1.0.2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpv559id48.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/strigops_habroptila/Strigops_habroptila.bStrHab1_v1.p.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp2tax9bgr.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/strix_occidentalis_caurina/Strix_occidentalis_caurina.Soccid_v01.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpwf8fb5rj.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/struthio_camelus_australis/Struthio_camelus_australis.ASM69896v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmprb4pntp9.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/suricata_suricatta/Suricata_suricatta.meerkat_22Aug2017_6uvM2_HiC.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpo9ek4jb4.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa/Sus_scrofa.Sscrofa11.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpf1moxrql.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_bamei/Sus_scrofa_bamei.Bamei_pig_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_berkshire/Sus_scrofa_berkshire.Berkshire_pig_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_gca007644095v1/Sus_scrofa_gca007644095v1.ASM764409v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_gca015776825v1/Sus_scrofa_gca015776825v1.Ninghe_Sus_1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_gca017957985v1/Sus_scrofa_gca017957985v1.ASM1795798v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_gca018555405v1/Sus_scrofa_gca018555405v1.ASM1855540v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_gca019290145v1/Sus_scrofa_gca019290145v1.CAU-K.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_gca020567905v1/Sus_scrofa_gca020567905v1.ASM2056790v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_gca021656055v1/Sus_scrofa_gca021656055v1.ASM2165605v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_gca024718415v1/Sus_scrofa_gca024718415v1.ASM2471841v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_hampshire/Sus_scrofa_hampshire.Hampshire_pig_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_jinhua/Sus_scrofa_jinhua.Jinhua_pig_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_landrace/Sus_scrofa_landrace.Landrace_pig_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_largewhite/Sus_scrofa_largewhite.Large_White_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_meishan/Sus_scrofa_meishan.Meishan_pig_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_pietrain/Sus_scrofa_pietrain.Pietrain_pig_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_rongchang/Sus_scrofa_rongchang.Rongchang_pig_v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_tibetan/Sus_scrofa_tibetan.Tibetan_Pig_v2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_usmarc/Sus_scrofa_usmarc.USMARCv1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/sus_scrofa_wuzhishan/Sus_scrofa_wuzhishan.minipig_v1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/taeniopygia_guttata/Taeniopygia_guttata.bTaeGut1_v1.p.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 321, in __download_gff_file\\n    f_gzip.write(data)\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 499, in func_wrapper\\n    return func(*args, **kwargs)\\n           ^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 517, in __exit__\\n    result = self.file.__exit__(exc, value, tb)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/takifugu_rubripes/Takifugu_rubripes.fTakRub1.2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpb9g7odna.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/terrapene_carolina_triunguis/Terrapene_carolina_triunguis.T_m_triunguis-2.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 234, in __init__\\n    genome_label = self.authority.value + \\\"-\\\" + taxon_id + \\\"-\\\" + genome_version\\n                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~\\nTypeError: can only concatenate str (not \\\"NoneType\\\") to str\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/tetraodon_nigroviridis/Tetraodon_nigroviridis.TETRAODON8.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp6ucx2fc9.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/theropithecus_gelada/Theropithecus_gelada.Tgel_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpfxlnac8u.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/tupaia_belangeri/Tupaia_belangeri.TREESHREW.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpt7zyzepu.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/tursiops_truncatus/Tursiops_truncatus.turTru1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpxzutj44w.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/urocitellus_parryii/Urocitellus_parryii.ASM342692v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpl33kals_.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ursus_americanus/Ursus_americanus.ASM334442v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmptpvbrj2w.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ursus_maritimus/Ursus_maritimus.UrsMar_1.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp2vgir2g5.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/ursus_thibetanus_thibetanus/Ursus_thibetanus_thibetanus.ASM966005v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpz1j_gd6x.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/varanus_komodoensis/Varanus_komodoensis.ASM479886v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpu7yw6tnx.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/vicugna_pacos/Vicugna_pacos.vicPac1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp87m6ep_2.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/vombatus_ursinus/Vombatus_ursinus.bare-nosed_wombat_genome_assembly.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp32fpod35.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/vulpes_vulpes/Vulpes_vulpes.VulVul2.2.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpw0htvk_m.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/xenopus_tropicalis/Xenopus_tropicalis.UCB_Xtro_10.0.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmp0hpsbua8.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/xiphophorus_couchianus/Xiphophorus_couchianus.Xiphophorus_couchianus-4.0.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpchsxoxmj.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/xiphophorus_maculatus/Xiphophorus_maculatus.X_maculatus-5.0-male.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpzrbv9r_o.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/zalophus_californianus/Zalophus_californianus.mZalCal1.pri.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpd7xfttql.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/zonotrichia_albicollis/Zonotrichia_albicollis.Zonotrichia_albicollis-1.0.1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpp3gmr39y.gz'\\n\",\n",
      "  \"https://ftp.ensembl.org/pub/release-114/gff3/zosterops_lateralis_melanops/Zosterops_lateralis_melanops.ASM128173v1.114.gff3.gz\": \"Traceback (most recent call last):\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 94, in scrape_ensembl\\n    result_path = serialize_bkbit_gff3(\\n                  ^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py\\\", line 122, in serialize_bkbit_gff3\\n    gff3 = gene_translator.Gff3.from_url(\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 274, in from_url\\n    return cls(\\n           ^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 238, in __init__\\n    self.gff_file, hash_values = self.__download_gff_file()\\n                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/site-packages/bkbit/data_translators/genome_annotation_translator.py\\\", line 302, in __download_gff_file\\n    with tempfile.NamedTemporaryFile(suffix=\\\".gz\\\", delete=False) as f_gzip:\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 582, in NamedTemporaryFile\\n    file = _io.open(dir, mode, buffering=buffering,\\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 579, in opener\\n    fd, name = _mkstemp_inner(dir, prefix, suffix, flags, output_type)\\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \\\"/Users/scott.daniel/miniconda3/envs/gene_mapper/lib/python3.12/tempfile.py\\\", line 256, in _mkstemp_inner\\n    fd = _os.open(file, flags, 0o600)\\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\nOSError: [Errno 28] No space left on device: '/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/tmpwmb8yz32.gz'\\n\"\n",
      "}\n",
      "SUCCESS\n",
      "that took 3.21e+01 minutes\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "[Errno 28] No space left on device: '../data/ensembl_download_failure.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mOSError\u001b[39m                                   Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[4]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m t0 = time.time()\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m ensembl_files_spec = \u001b[43mensembl_scraper\u001b[49m\u001b[43m.\u001b[49m\u001b[43mscrape_ensembl\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m      3\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdefault_ensembl_version\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m114\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      4\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdst_dir\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43m../data/ensembl_download\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      5\u001b[39m \u001b[43m    \u001b[49m\u001b[43mfailure_log_path\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43m../data/ensembl_download_failure.txt\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      6\u001b[39m \u001b[43m    \u001b[49m\u001b[43mspecific_files\u001b[49m\u001b[43m=\u001b[49m\u001b[43mbican_files\u001b[49m\n\u001b[32m      7\u001b[39m \u001b[43m)\u001b[49m\n\u001b[32m      8\u001b[39m dur = (time.time()-t0)/\u001b[32m60.0\u001b[39m\n\u001b[32m      9\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[33mscraping ENSEMBL took \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdur\u001b[38;5;132;01m:\u001b[39;00m\u001b[33m.2e\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m minutes\u001b[39m\u001b[33m'\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/KnowledgeEngineering/mmc_gene_mapper/src/mmc_gene_mapper/ensembl_download/scraper.py:110\u001b[39m, in \u001b[36mscrape_ensembl\u001b[39m\u001b[34m(default_ensembl_version, dst_dir, failure_log_path, specific_files, n_limit)\u001b[39m\n\u001b[32m    108\u001b[39m dur = (time.time()-t0)/\u001b[32m60.0\u001b[39m\n\u001b[32m    109\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[33mSUCCESS\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[33mthat took \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdur\u001b[38;5;132;01m:\u001b[39;00m\u001b[33m.2e\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m minutes\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m110\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mfailure_log_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mw\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m dst:\n\u001b[32m    111\u001b[39m     dst.write(json.dumps(failed_file_lookup))\n\u001b[32m    113\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m valid_files\n",
      "\u001b[31mOSError\u001b[39m: [Errno 28] No space left on device: '../data/ensembl_download_failure.txt'"
     ]
    }
   ],
   "source": [
    "t0 = time.time()\n",
    "ensembl_files_spec = ensembl_scraper.scrape_ensembl(\n",
    "    default_ensembl_version=114,\n",
    "    dst_dir='../data/ensembl_download',\n",
    "    failure_log_path='../data/ensembl_download_failure.txt',\n",
    "    specific_files=bican_files\n",
    ")\n",
    "dur = (time.time()-t0)/60.0\n",
    "print(f'scraping ENSEMBL took {dur:.2e} minutes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1a2e6fea-87ba-4ae0-b0a7-a6ba1ea884df",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ensembl_files_spec' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[43mensembl_files_spec\u001b[49m)\n",
      "\u001b[31mNameError\u001b[39m: name 'ensembl_files_spec' is not defined"
     ]
    }
   ],
   "source": [
    "ensembl_scraper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "720a2f1f-34d6-42c7-a29a-5a28002d3d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = pathlib.Path(\n",
    "    mmc_gene_mapper.__file__).parent.parent.parent / \"data\"\n",
    "assert data_dir.is_dir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd3b6d12-9a81-4585-9955-1343acb0fb05",
   "metadata": {},
   "outputs": [],
   "source": [
    "db_path = data_dir/\"gene_mapper_example.db\"\n",
    "assert db_path.parent.is_dir()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f02ee006-8ee1-4978-8d07-85d06dd3be8f",
   "metadata": {},
   "source": [
    "There are some files in\n",
    "```\n",
    "/allen/scratch/aibstemp/danielsf/gene_mapper_data\n",
    "```\n",
    "that are used to ingest data beyond what is available in the NCBI FTP server. Please copy those files to `../data/db_creation_data` in this respository (the cell below will fail if any of the files are not present)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06786610-8663-4821-9466-5db086caa52a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_file_spec=[\n",
    "    {\"type\": \"bkbit\",\n",
    "     \"path\": \"../data/db_creation_data/mouse_ENSEMBL-10090-98.jsonld\"},\n",
    "    {\"type\": \"bkbit\",\n",
    "     \"path\": \"../data/db_creation_data/human_ENSEMBL-9606-101.jsonld\"},\n",
    "    {\"type\": \"hmba_orthologs\",\n",
    "     \"path\": \"../data/db_creation_data/all_gene_ids.csv\",\n",
    "     \"name\": \"HMBA\",\n",
    "     \"baseline_species\": \"human\"}\n",
    "]\n",
    "\n",
    "error_msg = \"\"\n",
    "for spec in data_file_spec:\n",
    "    pth = pathlib.Path(spec['path'])\n",
    "    if not pth.is_file():\n",
    "        error_msg += f\"{pth} is not a file\\n\"\n",
    "if len(error_msg) > 0:\n",
    "    raise RuntimeError(error_msg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15278a0d-49a3-4942-8d12-dc52e366b70b",
   "metadata": {},
   "source": [
    "Below we instantiate the class that is used to actually do the mapping. The first time you run the cell, it will take ~ 30 minutes as it downloads ~ 4 GB of data from NCBI and ingests that data, alongside the local data from the cell above, to create an ~ 8 GB sqlite database that backs the mapping process.\n",
    "\n",
    "As long as you leave `clobber=False` below, you will not have to recreate that database on subsequent uses.\n",
    "\n",
    "**Note:** running this cell will also produce some warnings about genes that could not be ingested into the ortholog tables. These are due to genes in the ortholog table that have been deprecated by NCBI since the ortholog table (especially our by-hand HMBA ortholog table) was generated, meaning the gene mapper cannot connect them to species. We will have to think about how to handle versioning mismatches of this sort."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3123694f-af49-49e9-aa15-8360b10baf2a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gene_mapper = mapper.MMCGeneMapper(\n",
    "    db_path=db_path,\n",
    "    local_dir=data_dir,\n",
    "    data_file_spec=data_file_spec,\n",
    "    clobber=False,\n",
    "    force_download=False,\n",
    "    suppress_download_stdout=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f1a2daf-3fad-40bf-b5e2-2478c8caef22",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_mapper.db_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a5dede0-3fb4-4b04-88c4-b104659c7be5",
   "metadata": {},
   "source": [
    "## Contents of the database backing the mapper\n",
    "\n",
    "The database backing the `gene_mapper` contains the following tables\n",
    "\n",
    "### Metadata tables\n",
    "\n",
    "#### authority\n",
    "The `authority` table just tracks the gene identifying authorities the database knows about (i.e. ENSEMBL and NCBI). it consists of the columns\n",
    "\n",
    "- id -- an integer used for internal indexing\n",
    "- name -- the human-readable name of the authority\n",
    "\n",
    "#### citation\n",
    "The `citation` table tracks all of the datasets used to justify the various mappings (symbol-to-identifier, ENSEMBL-to-NCBI, orthologs, etc.). It consists of the columns\n",
    "\n",
    "- id -- an integer used for internal indexing\n",
    "- name -- a short, human-readable name for the citation (this will be used whenever you have to specify a citation when calling the `gene_mapper`'s methods\n",
    "- metadata -- a JSON serialized dict containing whatever information is necessary to recreate or specify the data backing this citation. This is a very free-form column as different entries will be specified in different ways (see below)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd687385-71d0-4a32-9511-5e8ec5dce2d1",
   "metadata": {},
   "source": [
    "We can list the available authorities like so"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aed469a-a7ef-4650-b502-e68731ce3166",
   "metadata": {},
   "outputs": [],
   "source": [
    "authority_list = gene_mapper.get_all_authorities()\n",
    "print(authority_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cc1581-fbb9-4716-9089-1b386d8a52ad",
   "metadata": {},
   "source": [
    "Similarly, we can list the available citations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7b988a9-1942-43d0-93f5-ae3f213fc766",
   "metadata": {},
   "outputs": [],
   "source": [
    "citation_list = gene_mapper.get_all_citations()\n",
    "print([c['name'] for c in citation_list])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a0ff8ea-e421-42aa-a4a3-debf3e2aae0a",
   "metadata": {},
   "source": [
    "As mentioned above, different citations contain different metadata. For instance, the `NCBI` citation (referring to the data downloaded from the NCBI FTP server) contains a list of the files downloaded their hashes, and the date they were downloaded on (NCBI updates its FTP server daily; it's not clear that they keep readily available version information around, though we can certainly investigate that). This is what the `NCBI` citation metadata looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "574f35ac-fc92-4ac2-9768-67c92c2d9089",
   "metadata": {},
   "outputs": [],
   "source": [
    "for citation in citation_list:\n",
    "    if citation['name'] == 'NCBI':\n",
    "        print(json.dumps(citation, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e33b8b7-145d-45ae-b3eb-b889bdd38ed2",
   "metadata": {},
   "source": [
    "The citations based on BICAN bkbit files, contain all of the GenomeAnnotation and GenomeAssembly information in those files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "022f5276-5b36-4242-bcc4-9d49c1ab0e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for citation in citation_list:\n",
    "    if citation['name'] == 'ENSEMBL-10090-98':\n",
    "        print(json.dumps(citation, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d2ddc70-ac65-47a8-97ed-1143f45dd532",
   "metadata": {},
   "source": [
    "The `HMBA` citation is an ingest of a lookup table created by our science team. As such, it has the least detailed information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07020636-8a9f-46ef-bd73-4915d1cdeb61",
   "metadata": {},
   "outputs": [],
   "source": [
    "for citation in citation_list:\n",
    "    if citation['name'] == 'HMBA':\n",
    "        print(json.dumps(citation, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8e9c8f1-3011-409c-9e7b-96eb1f1340e4",
   "metadata": {},
   "source": [
    "### Data tables\n",
    "\n",
    "The data tables in the database are\n",
    "\n",
    "#### NCBI_species\n",
    "The `NCBI_species` table just ingests the NCBI organism taxonomy and records\n",
    "- id -- the NCBI taxon ID of the species\n",
    "- name -- the human readable name of the species. **Note:** we record all possible names (\"common\", \"scientific\", etc.) of the species in different rows. That way, users can give us whatever name they have for a species and, if possible, we can return an `id` for cross-referencing with other tables.\n",
    "\n",
    "This is meant to be hidden from the user. The user specifies the species when performing a mapping and the gene_mapper makes the necessary database call. Below, we will make the call \"by hand\", so you can see what is in the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "168ce703-d762-4386-a812-0c58e491cd7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with sqlite3.connect(gene_mapper.db_path) as conn:\n",
    "    cursor = conn.cursor()\n",
    "    for name in ('house mouse', 'Mus musculus', 'mouse', 'human', 'Homo sapiens'):\n",
    "        result = cursor.execute(\n",
    "            \"\"\"\n",
    "            SELECT id FROM NCBI_species WHERE name=?\n",
    "            \"\"\",\n",
    "            (name,)\n",
    "        ).fetchall()\n",
    "        print(name,result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14fb7135-5d2a-4cdf-94e0-8bdb1739392e",
   "metadata": {},
   "source": [
    "**Note:** in the case of \"mouse\", there is more than one matching taxon. Calling a mapping with `species=\"mouse\"` will fail."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e2ad43d-c270-4d2d-8818-b0d22b7847d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import traceback\n",
    "\n",
    "mouse_symbols = [\"Xkr4\", \"Npbwr1\", \"not_a_symbol\", \"Rrs1\"]\n",
    "try:\n",
    "    gene_mapper.map_genes(\n",
    "        gene_list=mouse_symbols,\n",
    "        dst_species=\"mouse\",\n",
    "        dst_authority=\"ENSEMBL\"\n",
    "    )\n",
    "except Exception as err:\n",
    "    msg = traceback.format_exc()\n",
    "    print(\n",
    "        f\"Failed with error message:\\n=======\\n{msg}\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bedf5c05-1e66-4bf9-84f6-10853eda5c5e",
   "metadata": {},
   "source": [
    "A function is provided for listing all of the species the database knows about.\n",
    "\n",
    "**Note:** This is not a list of all of the species for which the database has gene information. It is merely a list of the species that can be associated with a taxon ID in the database (for cross-referencing with the tables that contain gene information)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d62030b-b2ad-4af7-a393-ab856a1eb164",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = gene_mapper.get_all_species()\n",
    "print(species_list[2000000:2000005])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05a3656f-1f18-4573-9fb8-fcbc9cac865c",
   "metadata": {},
   "outputs": [],
   "source": [
    "'Homo sapiens' in species_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae419412-4c1a-4a5b-beab-266fdf1ff983",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(species_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5631dbf9-e172-4bbc-9c08-77527c3722b0",
   "metadata": {},
   "source": [
    "#### gene\n",
    "The `gene` table contains all of the information about individual genes ingested into the database. Columns are\n",
    "- authority -- an integer for cross-referencing the `authority` table\n",
    "- citation -- an integer for cross-referencing the `citation` table\n",
    "- species_taxon -- an integer for cross-refereneing the `NCBI_species` table\n",
    "- id -- an integer for internal indexing\n",
    "- identifier -- a string. The full unique identifier (e.g. ENSMUS0000G12345) of the gene\n",
    "- symbol -- a string. The human-readable name of the gene (**Note:** the BICAN bkbit files contain `symbol` and `name` entries for genes. If these differ, both are ingested as a separate row in the `gene` table with `name` being recorded in `symbol` as appropriate).\n",
    "\n",
    "#### gene_equivalence\n",
    "The `gene_equivalence` table records mappings between authorities (ENSEMBL and NCBI). Its columns are\n",
    "- species_taxon -- an integer for cross-referencing with the `NCBI_species` table (in what species are we looking for equivalences?)\n",
    "- citation -- an integer for cross-referencing with the `citation` table (who said these genes were equivalent?)\n",
    "- authority0 -- an integer for cross-referencing with the `authority` table\n",
    "- gene0 -- an integer; the ID of the gene in `authority0`\n",
    "- authority1 -- an integer for cross-referncing with the `authority` table\n",
    "- gene1 -- an integer; the ID of the gene in `authority1` that is equivalent to `gene0` according to `citation`.\n",
    "\n",
    "**Note:** we record equivalences symmetrically, so that every `(gene0, gene1)` pair is also recorded as `(gene1, gene0)` so that users do not have to worry about which index (`gene0` or `gene1`) their input data is compared to.\n",
    "\n",
    "#### gene_ortholog\n",
    "The `gene_ortholog` table records cross-species ortholog relationships. Its columns are\n",
    "- authority -- an integer for cross-referencing the `authority` table (are we looking for NCBI orthologs or ENSEMBL orthologs?)\n",
    "- citation -- an integer for cross-referencing the `citation` table (who said these genes were orthologs?)\n",
    "- species -- an integer for cross-referencing the `NCBI_species` table\n",
    "- gene -- an integer, the ID of the gene in `species` whose orthologs we are looking for\n",
    "- ortholog_group -- an integer; genes with the same value of `ortholog_group` are orthologs of each other\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b06f1579-1a0b-498a-8533-a1349ef70543",
   "metadata": {},
   "source": [
    "## Actually mapping data\n",
    "\n",
    "The `gene_mapper` provides a single function to map gene identifiers from one form `(authority, species)` to another. To run it you specify\n",
    "- the list of gene identifeirs you have\n",
    "- the name of the species you want your genes mapped into (the mapper will automatically detect the species your genes are already in(\n",
    "- the authority of gene (ENSEMBL or NCBI) you want your genes mapped into\n",
    "- (optionally) the name of the citation to use for any cross species ortholog mappings (default is to use the NCBI lookup table)\n",
    "\n",
    "The mapper will then go through the necessary transformations and return a mapped list of gene identifiers for you."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b101728c-7ce8-4b0b-922a-26b6b721b8de",
   "metadata": {},
   "source": [
    "### Mapping symbols to identifiers\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e819cc2a-b66c-4d24-8e68-11d4e2cae8e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_symbols = [\"Xkr4\", \"Npbwr1\", \"not_a_symbol\", \"Rrs1\"]\n",
    "mouse_ens_mapping = gene_mapper.map_genes(\n",
    "    gene_list=mouse_symbols,\n",
    "    dst_species=\"Mus musculus\",\n",
    "    dst_authority=\"ENSEMBL\"\n",
    ")\n",
    "print(json.dumps(mouse_ens_mapping, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92c19413-d2db-4c95-87c9-1288c04d3344",
   "metadata": {},
   "source": [
    "For ease of reading, here is the output broken down into metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47c529fc-edbb-408d-a4f0-dd2f4075d716",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(json.dumps(mouse_ens_mapping['metadata'], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efeb945b-2532-404e-bdcd-041d9cbd657f",
   "metadata": {},
   "source": [
    "and actual mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7a8ffb2-e609-4c83-abe7-8d8536652850",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(json.dumps(mouse_ens_mapping['gene_list'], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "268dfc10-1ce4-42c4-928a-8caba10d5f77",
   "metadata": {},
   "source": [
    "Here is a similar call mapping symbols to identifiers in NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3016fa95-15a5-4cf4-b1d8-9b6031183b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_ncbi_mapping = gene_mapper.map_genes(\n",
    "    gene_list=mouse_symbols,\n",
    "    dst_species=\"Mus musculus\",\n",
    "    dst_authority=\"NCBI\"\n",
    ")\n",
    "print(json.dumps(mouse_ncbi_mapping, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b219d40-cffb-45de-9306-05d4166aaa72",
   "metadata": {},
   "source": [
    "### Mapping from ENSEMBL to NCBI\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ceb93a-2539-4d79-86c6-dbfe4f34ed7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_to_ncbi = gene_mapper.map_genes(\n",
    "    dst_authority='NCBI',\n",
    "    gene_list=[\"ENSMUSG00000051951\",\n",
    "               \"ENSMUSG00000030337\",\n",
    "               \"ENSMUSG00000087247\",\n",
    "               \"nope\",\n",
    "               \"ENSMUSG00000025911\"],\n",
    "    dst_species=\"Mus musculus\"\n",
    ")\n",
    "print(json.dumps(ens_to_ncbi, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58e593be-05dc-4df9-8fbd-9abd678b197f",
   "metadata": {},
   "source": [
    "Here is a call mapping from NCBI to ENSEMBL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbad7610-4142-45db-9f39-db97163c3280",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_to_ens = gene_mapper.map_genes(\n",
    "    dst_authority='ENSEMBL',\n",
    "    gene_list=[\"NCBIGene:67269\", \"NCBIGene:54200\", \"nope\", \"NCBIGene:70911\"],\n",
    "    dst_species=\"Mus musculus\"\n",
    ")\n",
    "print(json.dumps(ncbi_to_ens, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0de3aa5c-2ebe-4921-8d9b-b8851eff6220",
   "metadata": {},
   "source": [
    "To simulate \"production scale,\" we will now download all of the genes in the Yao et al. 2023 10X gene panel and try mappping them from ENSEMBL to NCBI to see how long it takes to process that many genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e77f845-8585-4bab-8f79-1455a09b7d38",
   "metadata": {},
   "outputs": [],
   "source": [
    "wmb_gene_df = abc_cache.get_metadata_dataframe(\n",
    "    directory='WMB-10X',\n",
    "    file_name='gene'\n",
    ")\n",
    "wmb_genes = wmb_gene_df.gene_identifier.values\n",
    "t0 = time.time()\n",
    "wmb_ncbi = gene_mapper.map_genes(\n",
    "    dst_authority='NCBI',\n",
    "    gene_list=wmb_genes,\n",
    "    dst_species='Mus musculus',\n",
    ")\n",
    "dur = time.time()-t0\n",
    "print(f'mapping {len(wmb_genes)} genes took {dur:.2e} seconds')\n",
    "\n",
    "n_mapped = 0\n",
    "for k in wmb_ncbi['gene_list']:\n",
    "    if 'UNMAPPABLE' not in k:\n",
    "        n_mapped += 1\n",
    "print(f'{n_mapped} genes had unique NCBI matches')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f2f4586-8ded-4785-abe4-81f89075e263",
   "metadata": {},
   "source": [
    "Now we will download the marker genes used by MapMyCells for the Whole Mouse Brain taxonomy and try mapping them from ENSEMBL to NCBI."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d377c657-9985-4bf6-8a73-8d7a4d19d235",
   "metadata": {},
   "outputs": [],
   "source": [
    "src_path = \"https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/mapmycells/WMB-10X/20240831/mouse_markers_230821.json\"\n",
    "wmb_marker_path = pathlib.Path('../data/mouse_markers_230821.json')\n",
    "if not wmb_marker_path.is_file():\n",
    "    print(f\"downloading {wmb_marker_path}\")\n",
    "    process = subprocess.Popen(\n",
    "        args=[\n",
    "           \"wget\",\n",
    "            src_path,\n",
    "            \"-O\",\n",
    "            str(wmb_marker_path),\n",
    "        ],\n",
    "        stderr=subprocess.DEVNULL\n",
    "    )\n",
    "    process.wait()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de440719-7e4b-49eb-898f-803a84215488",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes = set()\n",
    "with open(wmb_marker_path, 'rb') as src:\n",
    "    marker_lookup = json.load(src)\n",
    "for key in marker_lookup:\n",
    "    if key in ('log', 'metadata'):\n",
    "        continue\n",
    "    marker_genes = marker_genes.union(set(marker_lookup[key]))\n",
    "marker_genes = sorted(marker_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34db75d7-ca7f-4daa-9732-8729fcfab7c6",
   "metadata": {},
   "source": [
    "Let's now map our mouse marker genes from ENSEMBL to NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a611509d-1366-4cba-983f-4f54d5805b43",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'{len(marker_genes)} marker genes')\n",
    "t0 = time.time()\n",
    "marker_ncbi = gene_mapper.map_genes(\n",
    "    dst_authority='NCBI',\n",
    "    gene_list=marker_genes,\n",
    "    dst_species='Mus musculus'\n",
    ")\n",
    "dur = time.time()-t0\n",
    "print(f'mappiing {len(marker_genes)} genes took {dur:.2e} seconds')\n",
    "\n",
    "n_mapped = 0\n",
    "n_one_to_one = 0\n",
    "for k in marker_ncbi['gene_list']:\n",
    "    if 'UNMAPPABLE' not in k:\n",
    "        n_mapped += 1\n",
    "\n",
    "print(f'{n_mapped} genes had unique NCBI matches')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18007eeb-c0c7-4f9f-b6a2-ba6e3ae8a285",
   "metadata": {},
   "source": [
    "### Ortholog mapping\n",
    "\n",
    "To test ortholog mapping, let's take the genes from Yao et al. 2023 that only mapped to one NCBI gene and map them to human orthologs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f925d36c-603b-4d3d-bd7c-e585cd09d0ab",
   "metadata": {},
   "source": [
    "First let's just map 5 genes so that we can display the output and see what it looks like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6d5b0d1-353b-42fa-a0bf-6baf5eaac714",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_orthologs = gene_mapper.map_genes(\n",
    "    dst_authority='ENSEMBL',\n",
    "    dst_species='human',\n",
    "    gene_list=list(wmb_ncbi['gene_list'][10:15]) + ['not_an_actual_gene']\n",
    ")\n",
    "print(json.dumps(test_orthologs['gene_list'], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f91cd2c-6fee-40ef-95d3-42a700dde6b3",
   "metadata": {},
   "source": [
    "Now let's map all 24,000 genes and see how long that takes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ff5fd99-4d32-4a27-962a-6458c8116671",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40aacfd2-8a6a-47d7-bddc-c7a2f72a2c4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f70b0ce1-694d-42ea-907c-18f68d6ffc62",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "wmb_to_whb_orthologs = gene_mapper.map_genes(\n",
    "    dst_authority='ENSEMBL',\n",
    "    dst_species='human',\n",
    "    gene_list=wmb_ncbi['gene_list']\n",
    ")\n",
    "dur = time.time()-t0\n",
    "print(f'mapping {len(wmb_ncbi['gene_list'])} orthologs took {dur:.2e} seconds')\n",
    "n_mapped = 0\n",
    "for k in wmb_to_whb_orthologs['gene_list']:\n",
    "    if 'UNMAPPABLE' not in k:\n",
    "        n_mapped += 1\n",
    "print(f'{n_mapped} genes had orthologs')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cc34e39-8222-4552-8b0a-2f9c5efc9ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(json.dumps(wmb_to_whb_orthologs['metadata'], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ab77854-a424-417b-95fe-b6bd33d224f1",
   "metadata": {},
   "source": [
    "To map orthologs using the \"by-hand\" look up table created by our science team, change the kwarg `citation='HMBA'`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67551f79-568a-48ff-9f1e-3ce18676167e",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "wmb_to_whb_orthologs_hmba = gene_mapper.map_genes(\n",
    "    dst_authority='NCBI',\n",
    "    dst_species='human',\n",
    "    gene_list=wmb_ncbi['gene_list'],\n",
    "    ortholog_citation='HMBA'\n",
    ")\n",
    "dur = time.time()-t0\n",
    "print(f'mapping {len(wmb_ncbi[\"gene_list\"])} orthologs took {dur:.2e} seconds')\n",
    "n_mapped = 0\n",
    "for k in wmb_to_whb_orthologs_hmba['gene_list']:\n",
    "    if 'UNMAPPABLE' not in k:\n",
    "            n_mapped += 1\n",
    "print(f'{n_mapped} genes had orthologs')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24c6b414-f695-4f60-8129-ad8e043b2a08",
   "metadata": {},
   "source": [
    "#### Map from mouse to naked mole rat\n",
    "\n",
    "Now, inspired by a recent community forum post, let's take our mouse marker genes, convert them into NCBI identifiers, and finally convert those in to naked mole rat orthologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5177766-aacb-4f1c-8998-474530c890f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()\n",
    "\n",
    "naked_mole_rat = gene_mapper.map_genes(\n",
    "    dst_authority='ENSEMBL',\n",
    "    dst_species='naked mole rat',\n",
    "    gene_list=marker_genes,\n",
    "    ortholog_citation='NCBI'\n",
    ")\n",
    "\n",
    "dur = time.time()-t0\n",
    "print(f\"mapping took {dur:.2e} seconds\")\n",
    "n_mapped = 0\n",
    "for g in naked_mole_rat['gene_list']:\n",
    "    if 'UNMAPPABLE' not in g:\n",
    "        n_mapped += 1\n",
    "print(f\"{n_mapped} genes of {len(marker_genes)} had orthologs\")\n",
    "print(marker_genes[:5])\n",
    "print(naked_mole_rat['gene_list'][:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8773ade5-5760-48a5-a4b6-cc61d30f9666",
   "metadata": {},
   "source": [
    "Here is the metadata for the naked mole rat ortholog mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef0a81ca-e857-4896-affc-4e20dd202088",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(json.dumps(naked_mole_rat['metadata'], indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "620829de-0ca5-49af-9942-aa2cba055d2a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
